sql/colexec: TestExternalDistinctSpilling failed

sql/colexec.TestExternalDistinctSpilling [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3505599&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3505599&tab=artifacts#/) on master @ [b50be15e0d7aa15da242d0e3949e14d65b798010](https://github.com/cockroachdb/cockroach/commits/b50be15e0d7aa15da242d0e3949e14d65b798010):


```
=== RUN   TestExternalDistinctSpilling
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExternalDistinctSpilling075815020
    test_log_scope.go:80: use -show-logs to present logs inline
    external_distinct_test.go:224: 
        	Error Trace:	external_distinct_test.go:224
        	Error:      	Not equal: 
        	            	expected: 7
        	            	actual  : 6
        	Test:       	TestExternalDistinctSpilling
        	Messages:   	the spilling didn't occur in all cases
    panic.go:965: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExternalDistinctSpilling075815020
    panic.go:965: test-monitor: unexpected 1413120 leftover bytes
        goroutine 29 [running]:
        runtime/debug.Stack(0xc001245290, 0x303aa40, 0xc000f1a078)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x9f
        github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func1()
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:110 +0x27e
        panic(0x303aa40, 0xc000f1a078)
        	/usr/local/go/src/runtime/panic.go:965 +0x1b9
        github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic(0x449ab50, 0xc000072108, 0xc0015ec000, 0x32cd6a3, 0x20, 0xc001245530, 0x2, 0x2)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:342 +0x245
        github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc0001bfae0, 0x449ab50, 0xc000072108, 0xc00123ce01)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:412 +0x2ca
        github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:392
        github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*EvalContext).Stop(0xc000574000, 0x449ab50, 0xc000072108)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3663 +0x4b
        runtime.Goexit()
        	/usr/local/go/src/runtime/panic.go:613 +0x1e5
        testing.(*common).FailNow(0xc0003b8600)
        	/usr/local/go/src/testing/testing.go:741 +0x3c
        github.com/stretchr/testify/require.Equal(0x4470518, 0xc0003b8600, 0x2d91840, 0x5b0f658, 0x2d91840, 0x5b0f650, 0xc001245ab8, 0x1, 0x1)
        	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/stretchr/testify/require/require.go:165 +0x110
        github.com/cockroachdb/cockroach/pkg/sql/colexec.TestExternalDistinctSpilling(0xc0003b8600)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/external_distinct_test.go:224 +0xcd6
        testing.tRunner(0xc0003b8600, 0x381d848)
        	/usr/local/go/src/testing/testing.go:1193 +0xef
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1238 +0x2b3
--- FAIL: TestExternalDistinctSpilling (5.01s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestExternalDistinctSpilling PKG=./pkg/sql/colexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/sql-queries yuzefovich
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExternalDistinctSpilling.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/colexec.TestExternalDistinctSpilling [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3504945&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3504945&tab=artifacts#/) on master @ [e84a4844e6328955eaa0d913b8e9d047d64a39f1](https://github.com/cockroachdb/cockroach/commits/e84a4844e6328955eaa0d913b8e9d047d64a39f1):


```
=== RUN   TestExternalDistinctSpilling
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExternalDistinctSpilling202796765
    test_log_scope.go:80: use -show-logs to present logs inline
    external_distinct_test.go:224: 
        	Error Trace:	external_distinct_test.go:224
        	Error:      	Not equal: 
        	            	expected: 7
        	            	actual  : 6
        	Test:       	TestExternalDistinctSpilling
        	Messages:   	the spilling didn't occur in all cases
    panic.go:971: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExternalDistinctSpilling202796765
    panic.go:971: test-monitor: unexpected 1331200 leftover bytes
        goroutine 178 [running]:
        runtime/debug.Stack(0x9ecb698, 0x1, 0xc0010a6060)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xab
        github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func1()
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:110 +0xca
        panic(0x4865a40, 0xc0010a6060)
        	/usr/local/go/src/runtime/panic.go:971 +0x499
        github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic(0x79ae990, 0xc00019c010, 0xc000560000, 0x4afde53, 0x20, 0xc000543380, 0x2, 0x2)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:342 +0x445
        github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc00029bd60, 0x79ae990, 0xc00019c010, 0x1)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:412 +0x4cd
        github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:392
        github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*EvalContext).Stop(0xc000b20000, 0x79ae990, 0xc00019c010)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3663 +0x73
        runtime.Goexit()
        	/usr/local/go/src/runtime/panic.go:632 +0x146
        testing.(*common).FailNow(0xc00076f200)
        	/usr/local/go/src/testing/testing.go:741 +0x65
        github.com/stretchr/testify/require.Equal(0x7983e00, 0xc00076f200, 0x45ba700, 0x91b47f8, 0x45ba700, 0x91b47f0, 0xc000543a58, 0x1, 0x1)
        	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/stretchr/testify/require/require.go:165 +0x125
        github.com/cockroachdb/cockroach/pkg/sql/colexec.TestExternalDistinctSpilling(0xc00076f200)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/external_distinct_test.go:224 +0x1086
        testing.tRunner(0xc00076f200, 0x504e6f8)
        	/usr/local/go/src/testing/testing.go:1193 +0x203
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1238 +0x5d8
--- FAIL: TestExternalDistinctSpilling (16.74s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestExternalDistinctSpilling PKG=./pkg/sql/colexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/sql-queries yuzefovich
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExternalDistinctSpilling.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/colexec.TestExternalDistinctSpilling [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3505581&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3505581&tab=artifacts#/) on master @ [b50be15e0d7aa15da242d0e3949e14d65b798010](https://github.com/cockroachdb/cockroach/commits/b50be15e0d7aa15da242d0e3949e14d65b798010):


```
=== RUN   TestExternalDistinctSpilling
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExternalDistinctSpilling189174653
    test_log_scope.go:80: use -show-logs to present logs inline
    external_distinct_test.go:224: 
        	Error Trace:	external_distinct_test.go:224
        	Error:      	Not equal: 
        	            	expected: 5
        	            	actual  : 4
        	Test:       	TestExternalDistinctSpilling
        	Messages:   	the spilling didn't occur in all cases
    panic.go:971: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExternalDistinctSpilling189174653
    panic.go:971: test-monitor: unexpected 706560 leftover bytes
        goroutine 123 [running]:
        runtime/debug.Stack(0x9ecb698, 0x1, 0xc000ad46a8)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xab
        github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func1()
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:110 +0xca
        panic(0x4865a40, 0xc000ad46a8)
        	/usr/local/go/src/runtime/panic.go:971 +0x499
        github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic(0x79ae990, 0xc000074128, 0xc0003ce000, 0x4afde53, 0x20, 0xc00112d380, 0x2, 0x2)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:342 +0x445
        github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc0003fd180, 0x79ae990, 0xc000074128, 0x1)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:412 +0x4cd
        github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:392
        github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*EvalContext).Stop(0xc000408000, 0x79ae990, 0xc000074128)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3663 +0x73
        runtime.Goexit()
        	/usr/local/go/src/runtime/panic.go:632 +0x146
        testing.(*common).FailNow(0xc000001980)
        	/usr/local/go/src/testing/testing.go:741 +0x65
        github.com/stretchr/testify/require.Equal(0x7983e00, 0xc000001980, 0x45ba700, 0x91b47e8, 0x45ba700, 0x91b47e0, 0xc00110fa58, 0x1, 0x1)
        	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/stretchr/testify/require/require.go:165 +0x125
        github.com/cockroachdb/cockroach/pkg/sql/colexec.TestExternalDistinctSpilling(0xc000001980)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/external_distinct_test.go:224 +0x1086
        testing.tRunner(0xc000001980, 0x504e6f8)
        	/usr/local/go/src/testing/testing.go:1193 +0x203
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1238 +0x5d8
--- FAIL: TestExternalDistinctSpilling (1.02s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestExternalDistinctSpilling PKG=./pkg/sql/colexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/sql-queries yuzefovich
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExternalDistinctSpilling.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/colexec.TestExternalDistinctSpilling [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3506722&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3506722&tab=artifacts#/) on master @ [04f75fc4f0d3a65460a8266c41508fb0892e0e6e](https://github.com/cockroachdb/cockroach/commits/04f75fc4f0d3a65460a8266c41508fb0892e0e6e):


```
=== RUN   TestExternalDistinctSpilling
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExternalDistinctSpilling862024683
    test_log_scope.go:80: use -show-logs to present logs inline
    external_distinct_test.go:224: 
        	Error Trace:	external_distinct_test.go:224
        	Error:      	Not equal: 
        	            	expected: 5
        	            	actual  : 4
        	Test:       	TestExternalDistinctSpilling
        	Messages:   	the spilling didn't occur in all cases
    panic.go:965: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExternalDistinctSpilling862024683
    panic.go:965: test-monitor: unexpected 583680 leftover bytes
        goroutine 226 [running]:
        runtime/debug.Stack(0xc0010c9290, 0x303dda0, 0xc0004a6078)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x9f
        github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func1()
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:110 +0x27e
        panic(0x303dda0, 0xc0004a6078)
        	/usr/local/go/src/runtime/panic.go:965 +0x1b9
        github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic(0x449e6d0, 0xc000072108, 0xc0015f0000, 0x32d0d43, 0x20, 0xc0010c9530, 0x2, 0x2)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:342 +0x245
        github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc00013eb40, 0x449e6d0, 0xc000072108, 0xc00144eb01)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:412 +0x2ca
        github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:392
        github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*EvalContext).Stop(0xc000396000, 0x449e6d0, 0xc000072108)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3663 +0x4b
        runtime.Goexit()
        	/usr/local/go/src/runtime/panic.go:613 +0x1e5
        testing.(*common).FailNow(0xc0006ffb00)
        	/usr/local/go/src/testing/testing.go:741 +0x3c
        github.com/stretchr/testify/require.Equal(0x4474038, 0xc0006ffb00, 0x2d948c0, 0x5b14748, 0x2d948c0, 0x5b14740, 0xc0010c9ab8, 0x1, 0x1)
        	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/stretchr/testify/require/require.go:165 +0x110
        github.com/cockroachdb/cockroach/pkg/sql/colexec.TestExternalDistinctSpilling(0xc0006ffb00)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/external_distinct_test.go:224 +0xcd6
        testing.tRunner(0xc0006ffb00, 0x3821128)
        	/usr/local/go/src/testing/testing.go:1193 +0xef
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1238 +0x2b3
--- FAIL: TestExternalDistinctSpilling (1.12s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestExternalDistinctSpilling PKG=./pkg/sql/colexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/sql-queries yuzefovich
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExternalDistinctSpilling.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql/colexec.TestExternalDistinctSpilling [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3506448&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3506448&tab=artifacts#/) on master @ [aca7e0ee34b79f2dc502055d2b2a9c1982da79c2](https://github.com/cockroachdb/cockroach/commits/aca7e0ee34b79f2dc502055d2b2a9c1982da79c2):


```
=== RUN   TestExternalDistinctSpilling
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExternalDistinctSpilling982124321
    test_log_scope.go:80: use -show-logs to present logs inline
    external_distinct_test.go:224: 
        	Error Trace:	external_distinct_test.go:224
        	Error:      	Not equal: 
        	            	expected: 5
        	            	actual  : 4
        	Test:       	TestExternalDistinctSpilling
        	Messages:   	the spilling didn't occur in all cases
    panic.go:971: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExternalDistinctSpilling982124321
    panic.go:971: test-monitor: unexpected 583680 leftover bytes
        goroutine 119 [running]:
        runtime/debug.Stack(0x9ecb698, 0x1, 0xc000616060)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xab
        github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func1()
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:110 +0xca
        panic(0x4865a40, 0xc000616060)
        	/usr/local/go/src/runtime/panic.go:971 +0x499
        github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic(0x79ae990, 0xc0001a6010, 0xc0003c8000, 0x4afde53, 0x20, 0xc00013d380, 0x2, 0x2)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:342 +0x445
        github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop(0xc0001ffb80, 0x79ae990, 0xc0001a6010, 0x1)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:412 +0x4cd
        github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop(...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:392
        github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*EvalContext).Stop(0xc000212280, 0x79ae990, 0xc0001a6010)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3663 +0x73
        runtime.Goexit()
        	/usr/local/go/src/runtime/panic.go:632 +0x146
        testing.(*common).FailNow(0xc001124780)
        	/usr/local/go/src/testing/testing.go:741 +0x65
        github.com/stretchr/testify/require.Equal(0x7983e00, 0xc001124780, 0x45ba700, 0x91b47e8, 0x45ba700, 0x91b47e0, 0xc00013da58, 0x1, 0x1)
        	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/stretchr/testify/require/require.go:165 +0x125
        github.com/cockroachdb/cockroach/pkg/sql/colexec.TestExternalDistinctSpilling(0xc001124780)
        	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/external_distinct_test.go:224 +0x1086
        testing.tRunner(0xc001124780, 0x504e6f8)
        	/usr/local/go/src/testing/testing.go:1193 +0x203
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1238 +0x5d8
--- FAIL: TestExternalDistinctSpilling (2.00s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestExternalDistinctSpilling PKG=./pkg/sql/colexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/sql-queries yuzefovich
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExternalDistinctSpilling.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/colexec/external_distinct_test.go
+++ b/pkg/sql/colexec/external_distinct_test.go
@@ -221,7 +221,11 @@ func TestExternalDistinctSpilling(t *testing.T) {
 		require.Equal(t, 0, sem.GetCount(), "sem still reports open FDs at index %d", i)
 	}
 	if !spillingMightNotHappen {
-		require.Equal(t, numRuns, numSpills, "the spilling didn't occur in all cases")
+		// The "randomNullsInjection" subtest might not spill to disk when a
+		// large portion of rows is made NULL, so we allow two cases:
+		// - numSpills == numRuns
+		// - numSpills == numRuns - 1.
+		require.GreaterOrEqual(t, numSpills, numRuns-1, "the spilling didn't occur in all cases")
 	}
 
 	for _, acc := range accounts {

